### PR TITLE
lint for volume/*

### PR DESCRIPTION
**Pull request by MHBauer**
I am not familiar enough with volumes to understand totally what is going on in detail.

I need more expert eyes to make sure these are factual sentences.

for #14756

### CODE REVIEW

**Comment by calavera** [comment]
LGTM

**Comment by LK4D4** [comment]
Ping @cpuguy83 for review.

File: volume/drivers/api.go [hunk 1]
```diff
@@ -4,11 +4,13 @@ package volumedrivers
 
 import "github.com/docker/docker/volume"
 
+// NewVolumeDriver returns a driver has the given name mapped on the given client.
 func NewVolumeDriver(name string, c client) volume.Driver {
 	proxy := &volumeDriverProxy{c}
 	return &volumeDriverAdapter{name, proxy}
 }
 
+// VolumeDriver manages volumes from end to end.
```
Comment (cpuguy83) [inline volume/drivers/api.go]: VolumeDriver in this context is explicitly for facilitating communications with plugins, and that is all.

File: volume/drivers/extpoint.go [hunk 2]
```diff
@@ -43,6 +46,7 @@ func Unregister(name string) bool {
 	return true
 }
 
+// Lookup returns the driver associated with the given name.
```
Comment (cpuguy83) [inline volume/drivers/extpoint.go]: Lookup returns the volume driver associated with the given name. If the a driver with the given name has not been registered it checks if there is a VolumeDriver plugin available with the given name.

File: volume/local/local.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ const (
 
 var oldVfsDir = filepath.Join("vfs", "dir")
 
+// New looks at each directory in the given scope and adds it to the returned Root.
```
Comment (cpuguy83) [inline volume/local/local.go]: New instantiates a new Root instance with the provided scope. Scope is the base path to that the Root instance uses to store its volumes. The base path is created here if it does not exist.

File: volume/local/local.go [hunk 4]
```diff
@@ -51,6 +52,7 @@ func New(scope string) (*Root, error) {
 	return r, nil
 }
 
+// Root holds volumes associated with a scope
```
Comment (cpuguy83) [inline volume/local/local.go]: Root implements the Driver interface for the volume package and manages the creation/removal of volumes. It uses only standard vfs commands to create/remove dirs within its provided scope.

File: volume/local/local.go [hunk 5]
```diff
@@ -58,14 +60,18 @@ type Root struct {
 	volumes map[string]*Volume
 }
 
+// DataPath returns the constructed path of this volume.
 func (r *Root) DataPath(volumeName string) string {
 	return filepath.Join(r.path, volumeName, VolumeDataPathName)
 }
 
+// Name returns name of this (driver?)
```
Comment (cpuguy83) [inline volume/local/local.go]: Name returns the name of Root, defined in the volume package in the `DefaultDriverName` constant.

File: volume/local/local.go [hunk 6]
```diff
@@ -58,14 +60,18 @@ type Root struct {
...
 func (r *Root) Name() string {
-	return "local"
+	return volume.DefaultDriverName
 }
 
+// Create makes a local directory to associate with the returned
+// volume. Returns an error if the volume already exists.
```
Comment (cpuguy83) [inline volume/local/local.go]: Create creates a new `volume.Volume` with the provided name, creating the underlying directory tree required for this volume in the process.

File: volume/local/local.go [hunk 7]
```diff
@@ -90,6 +96,7 @@ func (r *Root) Create(name string) (volume.Volume, error) {
 	return v, nil
 }
 
+// Remove the directory with all contents
```
Comment (cpuguy83) [inline volume/local/local.go]: Remove removes the specified volume and all underlying data. If the given volume does not belong to this driver and error is returned. The volume is reference counted, if all all references are not released then the volume is not removed.

File: volume/local/local.go [hunk 8]
```diff
@@ -133,6 +140,7 @@ func (r *Root) scopedPath(realPath string) bool {
 	return false
 }
 
+// Volume holds a path associated with a unique name
```
Comment (cpuguy83) [inline volume/local/local.go]: Volume implements the `Volume` interface from the `volume` package and represents the volumes created by `Root`

File: volume/local/local.go [hunk 9]
```diff
@@ -144,22 +152,27 @@ type Volume struct {
...
 func (v *Volume) DriverName() string {
 	return v.driverName
 }
 
+// Path returns the data location.
 func (v *Volume) Path() string {
 	return v.path
 }
 
+// Mount returns the location where data is.
```
Comment (cpuguy83) [inline volume/local/local.go]: Mount returns the data location and is not really used for this driver, and is only here to satisfy the `Volume` interface.

File: volume/local/local.go [hunk 10]
```diff
@@ -144,22 +152,27 @@ type Volume struct {
...
 func (v *Volume) Mount() (string, error) {
 	return v.path, nil
 }
 
+// Unmount dissociates this local volume.
```
Comment (cpuguy83) [inline volume/local/local.go]: Umount is for satisfying the `Volume` interface and does not do anything in this driver.

File: volume/volume.go [hunk 11]
```diff
@@ -1,7 +1,9 @@
 package volume
 
-const DefaultDriverName = "local"
+// DefaultDriverName is the name for local owners of volumes.
+const DefaultDriverName string = "local"
```
Comment (cpuguy83) [inline volume/volume.go]: DefaultDriverName is the driver name used for the driver implemented in the `local` package.

**Comment by cpuguy83** [comment]
Some comments, these could probably even be improved.

File: volume/drivers/api.go (hunk 1 above)
Comment (MHBauer) [inline volume/drivers/api.go]: Looking at volumes and plugins for different reasons and I think I understand this a bit more. How is:
`VolumeDriver defines the available functions that volume plugins must implement.`

File: volume/local/local.go [hunk 12]
```diff
@@ -13,7 +16,7 @@ import (
 )
 
 // VolumeDataPathName is the name of the directory where the volume data is stored.
-// It uses a very distintive name to avoid colissions migrating data between
+// It uses a very distintive name to avoid colisions migrating data between
```
Comment (jfrazelle) [inline volume/local/local.go]: collisions

File: volume/local/local.go [hunk 13]
```diff
@@ -22,6 +25,9 @@ const (
 
 var oldVfsDir = filepath.Join("vfs", "dir")
 
+// New instantiates a new Root instance with the provided scope. Scope
+// is the base path to that the Root instance uses to store its
```
Comment (cpuguy83) [inline volume/local/local.go]: "base path to the Root instance"

File: hack/make/validate-lint [hunk 14]
```diff
@@ -39,6 +39,9 @@ packages=(
 	pkg/version
 	registry
 	utils
+	volume
+	volume/local
+	volume/volumedrivers
```
Comment (RichardScothern) [inline hack/make/validate-lint]: this should be volume/drivers

File: volume/local/local.go [hunk 15]
```diff
@@ -90,10 +104,14 @@ func (r *Root) Create(name string) (volume.Volume, error) {
 	return v, nil
 }
 
+// Remove removes the specified volume and all underlying data. If the
+// given volume does not belong to this driver and error is
```
Comment (RichardScothern) [inline volume/local/local.go]: 'an error'

File: volume/local/local.go [hunk 16]
```diff
@@ -133,7 +151,9 @@ func (r *Root) scopedPath(realPath string) bool {
 	return false
 }
 
-type Volume struct {
+// Volume implements the Volume interface from the volume package and
```
Comment (RichardScothern) [inline volume/local/local.go]: Volume -> localVolume

File: hack/make/validate-lint (hunk 14 above)
Comment (MHBauer) [inline hack/make/validate-lint]: I'm still learning go, so this is confusing to me. The directory path is `volume/drivers`, but the actual package name is `volumedrivers` which leads me to think it should be `volume/volumedrivers`.

**Comment by RichardScothern** [comment]
LGTM

**Comment by jfrazelle** [comment]
@cpuguy83 good for you?

**Comment by cpuguy83** [comment]
LGTM are we ok to merge with CI being wonky?

**Comment by jfrazelle** [comment]
ya :/ its stupid crap

**Comment by LK4D4** [comment]
I'm merging. At least it compiles.
LGTM

**Pull request merged**